Lib/fontTools/ttLib/tables/_g_l_y_f.py

```diff
@@ -254,57 +260,41 @@ def __len__(self):
 		assert len(self.glyphOrder) == len(self.glyphs)
 		return len(self.glyphs)
 
-	def getPhantomPoints(self, glyphName, ttFont, defaultVerticalOrigin=None):
+	def getPhantomPoints(self, glyphName, hMetrics, vMetrics=None):
 		"""Compute the four "phantom points" for the given glyph from its bounding box
 		and the horizontal and vertical advance widths and sidebearings stored in the
 		ttFont's "hmtx" and "vmtx" tables.
 
-		If the ttFont doesn't contain a "vmtx" table, the hhea.ascent is used as the
-		vertical origin, and the head.unitsPerEm as the vertical advance.
+		'hMetrics' should be ttFont['hmtx'].metrics.
 
-		The "defaultVerticalOrigin" (Optional[int]) is needed when the ttFont contains
-		neither a "vmtx" nor an "hhea" table, as may happen with 'sparse' masters.
-		The value should be the hhea.ascent of the default master.
+		'vMetrics' should be ttFont['vmtx'].metrics if there is "vmtx" or None otherwise.
+		If there is no vMetrics passed in, vertical phantom points are set to the zero coordinate.
 
 		https://docs.microsoft.com/en-us/typography/opentype/spec/tt_instructing_glyphs#phantoms
 		"""
 		glyph = self[glyphName]
-		assert glyphName in ttFont["hmtx"].metrics, ttFont["hmtx"].metrics
-		horizontalAdvanceWidth, leftSideBearing = ttFont["hmtx"].metrics[glyphName]
 		if not hasattr(glyph, 'xMin'):
 			glyph.recalcBounds(self)
+
+		horizontalAdvanceWidth, leftSideBearing = hMetrics[glyphName]
 		leftSideX = glyph.xMin - leftSideBearing
 		rightSideX = leftSideX + horizontalAdvanceWidth
-		if "vmtx" in ttFont:
-			verticalAdvanceWidth, topSideBearing = ttFont["vmtx"].metrics[glyphName]
+
+		if vMetrics:
+			verticalAdvanceWidth, topSideBearing = vMetrics[glyphName]
 			topSideY = topSideBearing + glyph.yMax
+			bottomSideY = topSideY - verticalAdvanceWidth
 		else:
-			# without vmtx, use ascent as vertical origin and UPEM as vertical advance
-			# like HarfBuzz does
-			verticalAdvanceWidth = ttFont["head"].unitsPerEm
-			if "hhea" in ttFont:
-				topSideY = ttFont["hhea"].ascent
-			else:
-				# sparse masters may not contain an hhea table; use the ascent
-				# of the default master as the vertical origin
-				if defaultVerticalOrigin is not None:
-					topSideY = defaultVerticalOrigin
```

Comment:
so you decided to remove this `defaultVerticalOrigin` parameter, why? what was wrong with it?

Reply:
If the font does not have `vmtx`, we were synthesizing glyph vertical origin (and advance-width). However, this is not necessary. If you look at either mutator or instancer, the deltas for vertical phantom points are simply ignored if font doesn't have `vmtx` table.

At rendering time, if font doesn't have `vmtx`, HarfBuzz, etc synthesize glyph vertical origin, from the intsance's `hhea`, which is already interpolated. Included deltas for vertical origin here will either be ignored (in which case we don't need them), or worse, applied on top of the synthesized vertical origin, which would be wrong.